src/NServiceBus.Core/HeaderDictionary.cs

```diff
@@ -0,0 +1,78 @@
+// unset
+
+#pragma warning disable 1591
+namespace NServiceBus
+{
+    using System.Collections.Generic;
+
+    /// <summary>
+    ///
+    /// </summary>
+    public class HeaderDictionary
+    {
+        IDictionary<string, string> incomingHeaders;
+
+        public HeaderDictionary()
+        {
+            this.incomingHeaders = new Dictionary<string, string>();
+        }
+
+        public HeaderDictionary(IDictionary<string, string> incomingHeaders)
+        {
+            this.incomingHeaders = incomingHeaders;
+        }
+
+        public HeaderDictionary(HeaderDictionary dictionary)
+        {
+            this.incomingHeaders = dictionary.GetDictionary();
+        }
+
+        IDictionary<string, string> GetDictionary()
+        {
+            return incomingHeaders;
+        }
+
+        IDictionary<string, string> GetOrCreateModifiable()
+        {
+            return incomingHeaders;
+        }
+
+        public IEnumerator<KeyValuePair<string, string>> GetEnumerator() => GetDictionary().GetEnumerator();
```

Comment:
would be interesting to see whether we could just remove the weird operations from the API here, e.g. `GetEnumerator`, `Clear`, `IsReadOnly`, etc. without anything breaking